src/foreign_key.c

```diff
@@ -0,0 +1,518 @@
+#include "foreign_key.h"
+#include "hypertable.h"
+
+static List *relation_get_referencing_fk(Oid reloid);
+static void constraint_get_trigger(Oid conoid, Oid *updtrigoid, Oid *deltrigoid);
+static char *ChooseForeignKeyConstraintNameAddition(int numkeys, AttrNumber *keys, Oid relid);
+static void createForeignKeyActionTriggers(Form_pg_constraint fk, Oid relid, Oid refRelOid,
+										   Oid constraintOid, Oid indexOid, Oid parentDelTrigger,
+										   Oid parentUpdTrigger);
+static void clone_constraint_on_chunk(const Chunk *chunk, Relation parentRel, Form_pg_constraint fk,
+									  int numfks, AttrNumber *conkey, AttrNumber *confkey,
+									  Oid *conpfeqop, Oid *conppeqop, Oid *conffeqop,
+#if PG15_GE
+									  int numfkdelsetcols, AttrNumber *confdelsetcols,
+#endif
+									  Oid parentDelTrigger, Oid parentUpdTrigger);
+
+static void
+propagate_fk(Relation ht_rel, HeapTuple fk_tuple, List *chunks)
+{
+	Form_pg_constraint fk = (Form_pg_constraint) GETSTRUCT(fk_tuple);
+
+	int numfks;
+	AttrNumber conkey[INDEX_MAX_KEYS];
+	AttrNumber confkey[INDEX_MAX_KEYS];
+	Oid conpfeqop[INDEX_MAX_KEYS];
+	Oid conppeqop[INDEX_MAX_KEYS];
+	Oid conffeqop[INDEX_MAX_KEYS];
+#if PG15_GE
+	int numfkdelsetcols;
+	AttrNumber confdelsetcols[INDEX_MAX_KEYS];
+#endif
+
+	DeconstructFkConstraintRow(fk_tuple,
+							   &numfks,
+							   conkey,
+							   confkey,
+							   conpfeqop,
+							   conppeqop,
+							   conffeqop
+#if PG15_GE
+							   ,
+							   &numfkdelsetcols,
+							   confdelsetcols
+#endif
+	);
+
+	Oid parentDelTrigger, parentUpdTrigger;
+	constraint_get_trigger(fk->oid, &parentUpdTrigger, &parentDelTrigger);
+
+	ListCell *lc;
+	foreach (lc, chunks)
+	{
+		Chunk *chunk = lfirst(lc);
+		clone_constraint_on_chunk(chunk,
+								  ht_rel,
+								  fk,
+								  numfks,
+								  conkey,
+								  confkey,
+								  conpfeqop,
+								  conppeqop,
+								  conffeqop,
+#if PG15_GE
+								  numfkdelsetcols,
+								  confdelsetcols,
+#endif
+								  parentDelTrigger,
+								  parentUpdTrigger);
+	}
+}
+
+/*
+ * Copy foreign key constraints from the main table to a chunk.
+ */
+void
+ts_chunk_copy_referencing_fk(const Hypertable *ht, const Chunk *chunk)
+{
+	ListCell *lc;
+	List *chunks = list_make1((Chunk *) chunk);
+	List *fks = relation_get_referencing_fk(ht->main_table_relid);
+
+	Relation ht_rel = table_open(ht->main_table_relid, AccessShareLock);
+	foreach (lc, fks)
+	{
+		HeapTuple fk_tuple = lfirst(lc);
+		propagate_fk(ht_rel, fk_tuple, chunks);
+	}
+	table_close(ht_rel, NoLock);
+}
+
+void
+ts_fk_propagate(Constraint *c, Hypertable *ht)
+{
+	ListCell *lc1;
+	ListCell *lc2;
+
+	List *fks = relation_get_referencing_fk(ht->main_table_relid);
+	List *chunks = ts_chunk_get_by_hypertable_id(ht->fd.id);
+
+	Relation parentRel = table_open(ht->main_table_relid, AccessShareLock);
+
+	foreach (lc1, fks)
+	{
+		HeapTuple tuple = lfirst(lc1);
+		Form_pg_constraint fk = (Form_pg_constraint) GETSTRUCT(tuple);
```

Comment:
This part looks a lot like propagate_fk function?

Reply:
did some cleanup here